tests/pthread/test_pthread_dispatch_to_thread.c

```diff
@@ -0,0 +1,290 @@
+#include <assert.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "emscripten.h"
+#include "emscripten/threading.h"
+
+// Disable leak checking since we have allocations that deliberately outlive
+// `main`.
+const char* __asan_default_options() { return "detect_leaks=0"; }
+
+pthread_t main_thread;
+
+typedef struct widget {
+  // `val` will be stored to `out` and the current thread will be stored to
+  // `thread` when the widget is run.
+  int* out;
+  int val;
+  pthread_t thread;
+
+  // Synchronization to allow waiting on a widget to run.
+  pthread_mutex_t mutex;
+  pthread_cond_t cond;
+
+  // Nonzero iff the widget has been run.
+  int done;
+
+  // Only used for async_as_sync tests.
+  em_async_as_sync_ctx* ctx;
+} widget;
+
+void init_widget(widget* w, int* out, int val) {
+  widget empty = {out, val};
+  *w = empty;
+  pthread_mutex_init(&w->mutex, NULL);
+  pthread_cond_init(&w->cond, NULL);
+}
+
+void destroy_widget(widget* w) {
+  pthread_mutex_destroy(&w->mutex);
+  pthread_cond_destroy(&w->cond);
+}
+
+widget* malloc_widget(int* out, int val) {
+  widget* w = malloc(sizeof(widget));
+  assert(w);
+  init_widget(w, out, val);
+  return w;
+}
+
+void free_widget(widget* w) {
+  destroy_widget(w);
+  free(w);
+}
+
+void run_widget(widget* w) {
+  printf("running widget %d on %s\n",
+         w->val,
+         pthread_equal(pthread_self(), main_thread) ? "main" : "pthread");
+  pthread_mutex_lock(&w->mutex);
+  if (w->out) {
+    *w->out = w->val;
+  }
+  w->thread = pthread_self();
+  w->done = 1;
+  pthread_mutex_unlock(&w->mutex);
+  pthread_cond_broadcast(&w->cond);
+}
+
+void await_widget(widget* w) {
+  pthread_mutex_lock(&w->mutex);
+  while (!w->done) {
+    pthread_cond_wait(&w->cond, &w->mutex);
+  }
+  pthread_mutex_unlock(&w->mutex);
+}
+
+// Helper functions we will dispatch to perform our work.
+
+void do_run_widget(void* arg) { run_widget((widget*)arg); }
+
+void run_and_free_widget(void* arg) {
+  widget* w = (widget*)arg;
+  run_widget(w);
+  free_widget(w);
+}
+
+void finish_running_widget(void* arg) {
+  widget* w = (widget*)arg;
+  run_widget(w);
+  emscripten_async_as_sync_finish(w->ctx);
+}
+
+void start_running_widget(em_async_as_sync_ctx* ctx, void* arg) {
+  ((widget*)arg)->ctx = ctx;
+  emscripten_async_call(finish_running_widget, arg, 0);
+}
+
+void start_and_finish_running_widget(em_async_as_sync_ctx* ctx, void* arg) {
+  ((widget*)arg)->ctx = ctx;
+  finish_running_widget(arg);
```

Comment:
This doesn't actually do an async operation on the called thread. How about using `emscripten_async_call()`?